### PR TITLE
Add "master_fill_policy" tag property and its support in "tile" and "corner" layouts

### PR DESCRIPTION
**Pull request by actionless**
Follow-up on https://github.com/awesomeWM/awesome/issues/298

myimage

### CODE REVIEW

**Comment by actionless** [comment]
i am not sure what it can be applied nicely to the rest of layouts

**Comment by actionless** [comment]
i just came up this the idea how it can be generalized:
instead of `arrange` of layout [1] call custom "arranger" if :
1) there are only master client(s) on tag
AND
2) master_fill_policy set to mwfact
AND
3) layout is not `floating` or `max`

but i'm not sure if it's better than now

1) https://github.com/awesomeWM/awesome/blob/master/lib/awful/layout/init.lua#L138

File: lib/awful/tag.lua [hunk 1]
```diff
@@ -409,6 +409,36 @@ function tag.getgap(t)
     return tag.getproperty(t, "useless_gap") or beautiful.useless_gap or 0
 end
 
+--- Set size fill policy for the master client(s)
+-- @tparam string policy Can be set to
+-- "expand" (fill all the available workarea) or
+-- "mwfact" (fill only an area inside the master width factor)
+-- @tparam tag t The tag to modify, if null tag.selected() is used.
```
Comment (blueyed) [inline lib/awful/tag.lua]: You could write this as `@tparam[opt=tag.selected()] tag The tag to modify`.

JFI, it's OK like it is.

File: lib/awful/tag.lua [hunk 2]
```diff
@@ -409,6 +409,36 @@ function tag.getgap(t)
...
+function tag.setmfpol(enabled, t)
+    local t = t or tag.selected()
+    tag.setproperty(t, "master_fill_policy", enabled)
+end
+
+--- Toggle size fill policy for the master client(s)
+-- @tparam tag t The tag to modify, if null tag.selected() is used.
+function tag.togglemfpol(t)
```
Comment (blueyed) [inline lib/awful/tag.lua]: This should document what it toggles between.

File: lib/awful/tag.lua [hunk 3]
```diff
@@ -409,6 +409,36 @@ function tag.getgap(t)
     return tag.getproperty(t, "useless_gap") or beautiful.useless_gap or 0
 end
 
+--- Set size fill policy for the master client(s)
+-- @tparam string policy Can be set to
+-- "expand" (fill all the available workarea) or
+-- "mwfact" (fill only an area inside the master width factor)
+-- @tparam tag t The tag to modify, if null tag.selected() is used.
+function tag.setmfpol(enabled, t)
```
Comment (blueyed) [inline lib/awful/tag.lua]: The argument name is misleading.. indicates to expect a boolean.

File: lib/awful/tag.lua [hunk 4]
```diff
@@ -409,6 +409,36 @@ function tag.getgap(t)
...
+    if tag.getmfpol(t) == "expand" then
+        tag.setmfpol("mwfact", t)
+    else
+        tag.setmfpol("expand", t)
+    end
+end
+
+--- Get size fill policy for the master client(s)
+-- @tparam tag t Optional tag.
+-- @treturn string Can be set to
```
Comment (blueyed) [inline lib/awful/tag.lua]: s/Can bet set to/Possible values are/

And also mention the default of `expand` here.

**Comment by blueyed** [comment]
Cool feature!

I think that a numerical value would be useful maybe, too.
This would behave like the "mwfact" mode, but can be different from it then.
But I see that the main benefit of using "mwfact" is that clients are not getting resized.

File: lib/awful/tag.lua (hunk 3 above)
Comment (actionless) [inline lib/awful/tag.lua]: oops, i did not renamed it from the time when it actually was a bool

**Comment by actionless** [comment]
>I think that a numerical value would be useful maybe, too.
This would behave like the "mwfact" mode, but can be different from it then.
But I see that the main benefit of using "mwfact" is that clients are not getting resized.

I did it intentionally because i wanted master to always have the same width.
But bright thing here is that what it's possible to add another policy which would behave differently or if it's set to number instead of string it could work instead of mwfact value (or that was what you did suggested?).

And i still need some feedback on generalization (https://github.com/awesomeWM/awesome/pull/323#issuecomment-122023593) and one more layout policy (https://github.com/awesomeWM/awesome/pull/323#issuecomment-121923384).

Btw, what about naming, should i change "mfpol" to smth different?

**Comment by blueyed** [comment]
> if it's set to number instead of string it could work instead of mwfact value (or that was what you did suggested?).

Yes.

> Btw, what about naming, should i change "mfpol" to smth different?

I'd like a more verbose name, but I see that it's based on the current naming scheme.  As long as the docs are good, I think it's OK.

**Comment by blueyed** [comment]
> And i still need some feedback on generalization (#323 (comment)) and one more layout policy (#323 (comment)).

Sorry, I cannot provide feedback for this really.. :/

**Comment by blueyed** [comment]
@actionless 
Do you think this is good as it is for master now? (given that you would like to get some feedback still)

**Comment by actionless** [comment]
i am using this option daily and don't see any issues with it

additional options can be added in a separate PR though

**Comment by blueyed** [comment]
Merged, thanks a lot!

**Pull request merged**